### PR TITLE
DEV: scipy.interpolate b-splines (periodic case)

### PR DESCRIPTION
**Pull request by diana-sukhoverkhova**
#### What does this implement/fix?
We developed and implemented an algorithm for computing coefficients for B-splines of degree ``k`` in periodic case.  The algorithm builds a matrix with ``k`` diagonals and upper right and lower left blocks of size ``(k-1)/2 x (k-1)/2`` (for now ``k`` is assumed to be odd). This matrix is stored in a form suitable for ``scipy.linalg.solve_banded``.  Due to the special form of matrix, it is possible to use Woodbury formula to optimize the computations. 

#### Additional information
In periodic case, ordinates and values of the first ``k-1`` derivatives are equal at the ends.

### CODE REVIEW

File: scipy/interpolate/_bsplines.py [hunk 1]
```diff
@@ -606,6 +607,174 @@ def _process_deriv_spec(deriv):
...
+        B-spline degree.
+        
+    Returns
+    -------
+    c : 1-D array, shape(n,)
+        Solution of the original SLE.
+        
+    Notes
+    -----
+    SLE - system of linear equations.
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:637]: Leftover line? Can go.

File: scipy/interpolate/_bsplines.py [hunk 2]
```diff
@@ -606,6 +607,174 @@ def _process_deriv_spec(deriv):
...
+    
+    'n' should be greater than 'k', otherwise corner block
+    elements will intersect with diagonals.
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:640]: Should be checked probably.

File: scipy/interpolate/_bsplines.py [hunk 3]
```diff
@@ -606,6 +607,174 @@ def _process_deriv_spec(deriv):
...
+
+    Almost diagonal matrix with upper right and lower left
+    blocks is solved with the following steps: blocks are
+    represented as vectors and new systems of linear equations
+    are constructed. Left parts of systems are the same - the
+    diagonal elements of the original matrix which makes it
+    possible to use a solution for banded matrices 
+    (``solve_banded``). Then the solution for the original SLE
+    is finally computed.
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:649]: Can probably add the matrix structure in ASCII pseudocode here.

File: scipy/interpolate/_bsplines.py [hunk 4]
```diff
@@ -606,6 +607,174 @@ def _process_deriv_spec(deriv):
...
+    n = A.shape[1] + 1
+
+    U = np.zeros((n - 1, k - 1))
+    V = np.zeros((k - 1, n - 1))  # V transpose 
+
+    # upper right
+
+    U[:bs, :bs] = ur
+    for j in range(bs): 
+        V[j, -bs + j] = 1
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:667]: minor: can be done with fancy indexing, `V[np.arange(bs), np.arange(bs) - bs]` or some such.

File: scipy/interpolate/_bsplines.py [hunk 5]
```diff
@@ -606,6 +607,174 @@ def _process_deriv_spec(deriv):
...
+    are constructed. Left parts of systems are the same - the
+    diagonal elements of the original matrix which makes it
+    possible to use a solution for banded matrices 
+    (``solve_banded``). Then the solution for the original SLE
+    is finally computed.
+
+    References
+    ----------
+    .. [1] William H. Press, Saul A. Teukolsky, William T. Vetterling
+           and Brian P. Flannery, Numerical Recipes, 2007, 2.7.3
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:654]: Sec 2.7.3 ?

File: scipy/interpolate/_bsplines.py [hunk 6]
```diff
@@ -606,6 +607,174 @@ def _process_deriv_spec(deriv):
         ords, vals = [], []
     return np.atleast_1d(ords, vals)
 
+def _woodbury_algorithm(A, ur, ll, b, k):
+    '''
+    Implementation of Woodbury algorithm applied to banded
+    matrices with two blocks in upper right and lower left
+    corners.
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:614]: "Solve a cyclic banded linear system with upper right and lower blocks of size `k` using the Woodbury formula" --- is this a fair description of this routine?

Then ASCII pseudocode structure of the matrix would help after the sentence of in the Notes section.

File: scipy/interpolate/_bsplines.py [hunk 7]
```diff
@@ -606,6 +607,174 @@ def _process_deriv_spec(deriv):
...
+        Z = np.concatenate((Z, zi), axis=0)
+
+    H = solve(np.identity(k - 1) + V @ Z.T, np.identity(k - 1))
+
+    y = solve_banded((bs, bs), A, b)
+    c = y - Z.T @ (H @ (V @ y))
+
+    return c
+
+def _periodic_nodes(x,l=1,r=1):
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:690]: nit: these are periodic knots, right?

Comment (ev-br) [inline scipy/interpolate/_bsplines.py:690]: space after comma please, here and elsewhere (until the linter CI job is happy)

File: scipy/interpolate/_bsplines.py [hunk 8]
```diff
@@ -606,6 +607,174 @@ def _process_deriv_spec(deriv):
...
+
+    For now, this function works only for odd ``k``.
+
+    '''
+    if k % 2 == 0:
+        raise NotImplementedError("Even k periodic case is not implemented yet.")    
+    n = y.shape[0]
+
+    if n <= k:
+        raise ValueError("n should be greater than k to form system.")
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:748]: Nit: ... which system? Consider reformulating. E.g., need at least k data point to fit a spline of degree k

File: scipy/interpolate/_bsplines.py [hunk 9]
```diff
@@ -606,6 +607,174 @@ def _process_deriv_spec(deriv):
...
+    
+    # solving periodic case using Woodbury formula
+    # set up RHS
+    A = np.zeros((k, n - 1)) # matrix of diagonals suitable for 'solve_banded'
+    for i in range(n-1):
+        A[:,i] = _bspl.evaluate_all_bspl(t, k, x[i], i + k)[:-1][::-1]
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:754]: [::-1][::-1] looks unusual, is it really needed?

File: scipy/interpolate/_bsplines.py [hunk 10]
```diff
@@ -606,6 +607,174 @@ def _process_deriv_spec(deriv):
...
+    offset = int((k-1)/2)
+    # upper right and lower left blocks of the original matrix
+    ur = np.zeros((offset,offset))
+    ll = np.zeros((offset,offset))
+    for i in range(1,offset + 1):
+        A[offset - i] = np.roll(A[offset - i],i)
+        if k % 2 == 1 or i < offset:
+            A[offset + i] = np.roll(A[offset + i], -i)
+            ur[-i:, i - 1] = np.copy(A[offset + i, -i:])
+        ll[-i, :i] = np.copy(A[offset - i, :i])
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:764]: certainly need some more docs to follow this part.

Maybe also consider writing explicit loops and move them to cython, like it's done for other boundary conditions.

File: scipy/interpolate/tests/test_bsplines.py [hunk 11]
```diff
@@ -1094,59 +1151,6 @@ def make_interp_full_matr(x, y, t, k):
     return c
 
 
-### XXX: 'periodic' interp spline using full matrices
```
Comment (ev-br) [inline scipy/interpolate/tests/test_bsplines.py:1097]: Why removing this? Maybe better keep this and test your implementation against.
(if this version is buggy, fix it)

**Review by ev-br — CHANGES_REQUESTED** [review]
Initial cursory review. Did not check the algorithmic part, also `make_interp_spline` interface. Will do separately.

Full disclosure: Egor and Diana work with me at HSE, so an additional review from someone else would be nice.

**Comment by ev-br** [comment]
When revising this, please also add tests for the number of points smaller than and equal to k, to catch edge cases similar to k=3 CubicSpline.

**Comment by diana-sukhoverkhova** [comment]
Now this implementation works with B-splines of even degree ```k```. We changed default knot vector (in Greville-like way) and added dependency on parity to construction of matrices.

File: scipy/interpolate/_bsplines.py [hunk 12]
```diff
@@ -606,6 +607,195 @@ def _process_deriv_spec(deriv):
...
+    '''
+    n = y.shape[0]
+
+    if n <= k:
+        raise ValueError("Need at least k + 1 data points to fit a spline "
+                         "of degree k.")
+    
+    nt = len(t) - k - 1
+
+    # size of block elements
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:763]: Consider reformulating

File: scipy/interpolate/_bsplines.py [hunk 13]
```diff
@@ -606,6 +607,195 @@ def _process_deriv_spec(deriv):
...
+    # upper right and lower left blocks
+    ur = np.zeros((kul, kul))
+    ll = np.zeros_like(ur)
+    
+    # `offset` is made to shift all the non-zero elements to the end of the
+    # matrix
+    _bspl._colloc(x, t, k, ab, offset=k)
+    
+    # remove zeros before the matrix
+    ab = ab[-k - (k + 1) % 2:, :]
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:778]: Why? Consider re formulating (end of the matrix? Before the matrix?)

File: scipy/interpolate/_bsplines.py [hunk 14]
```diff
@@ -606,6 +607,195 @@ def _process_deriv_spec(deriv):
...
+    
+    # The least elements in rows (except repetitions) are diagonals
+    # of block matricies. Upper right matrix is an upper triangular
+    # matrix while lower left is a lower triangular one.
+    for i in range(kul):
+        ur += np.diag(ab[-i - 1, i: kul], k=i)
+        ll += np.diag(ab[i, -kul - (k % 2): n - 1 + 2 * kul - i], k=-i)
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:785]: This is due to the special choice of the knot vector, correct?

File: scipy/interpolate/_bsplines.py [hunk 15]
```diff
@@ -606,6 +607,195 @@ def _process_deriv_spec(deriv):
...
+
+    # remove elements that occure in the last point
+    # (first and last points are equivalent)
+    A = ab[:, kul: -k + kul]
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:789]: Typo: occurE

File: scipy/interpolate/_bsplines.py [hunk 16]
```diff
@@ -736,37 +943,46 @@ def make_interp_spline(x, y, k=3, t=None, bc_type=None, axis=0,
...
 
+    x = _as_float_array(x, check_finite)
+    y = _as_float_array(y, check_finite)
+
+    y = np.rollaxis(y, axis)    # now internally interp axis is zero
+
+    if bc_type == 'periodic' and not np.allclose(y[0], y[-1], atol=1e-15):
+        raise ValueError("First and last points does not match while "
+                         "periodic case expected")
+
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:954]: Identical to what CubcSpline does?

File: scipy/interpolate/tests/test_bsplines.py [hunk 17]
```diff
@@ -819,6 +820,93 @@ def test_not_a_knot(self):
...
 
+    @pytest.mark.parametrize('k', [2, 3, 4, 5])
+    def test_periodic(self, k):
+        b = make_interp_spline(self.xx, self.yy, k=k, bc_type='periodic')
+        assert_allclose(b(self.xx), self.yy, atol=1e-14, rtol=1e-14)
+        # in periodic case it is expected equality of k-1 first
+        # derivatives at the boundaries
+        for i in range(k):
+            assert_allclose(b(self.xx[0], nu=i), b(self.xx[-1], nu=i), atol=1e-11)
+        # tests for axis=-1
```
Comment (ev-br) [inline scipy/interpolate/tests/test_bsplines.py:831]: How is this different from what's above?

File: scipy/interpolate/tests/test_bsplines.py [hunk 18]
```diff
@@ -819,6 +820,93 @@ def test_not_a_knot(self):
...
+            y[0] = y[-1]
+            assert_raises(ValueError, make_interp_spline, x, y, k, None,
+            'periodic')
+
+        # first and last points should match when periodic case expected
+        n = 8
+        x = np.sort(np.random.random_sample(n))
+        y = np.random.random_sample(n)
+        y[0] = y[-1] - 1  # to be sure that they are not equal
+        assert_raises(ValueError, make_interp_spline, x, y, k, None,
```
Comment (ev-br) [inline scipy/interpolate/tests/test_bsplines.py:868]: Use 
```
with assert_raises(ValueError) :
    make_interp_spline(...) 
```

**Review by ev-br — CHANGES_REQUESTED** [review]
It's getting closer. One more review round, algo review still pending.

**Review by ev-br — APPROVED** [review]
Looks good modulo small nits, which can taken care of in a sequel or not at all. 

Am planning to merge it soon (tm), unless there are more comments --- Would be nice to have a second pair of eyes on this BTW.

**Comment by tylerjereddy** [comment]
> Looks good modulo small nits, which can taken care of in a sequel or not at all.

Alright, given that the only CI failure is unrelated I'm basically relying on this assessment and my own quick scan through the code for sanity checking and I'm going to merge it. It looks like @ev-br commented on an old test being deleted, but then resolved that conversation, so I'm assuming that is "ok"/approved here.

> Full disclosure: Egor and Diana work with me at HSE, so an additional review from someone else would be nice.

Agreed, but fairly specialized stuff, and we do have folks that appear to work at the same organization merging each others PRs without too much fuss already. Nice when we can get another person to check of course.

**Comment by tylerjereddy** [comment]
Thanks @diana-sukhoverkhova @ev-br

**Pull request merged**